### ISSUE TITLE
Make it harder to avoid a virtual environment

### ISSUE THREAD

**Issue by dimbleby** [python-poetry/poetry#6398]
Every time someone comes here with something going wrong after they set `poetry config virtualenvs.create false` the answer is "oh, you didn't want to do that".  It's a footgun.

Can anything be done?

- update the docs with more cautions?
- make that option harder to use / print a big old warning when it is used?
- add a new option with meaning "I really do want to use the system environment and acknowledge that it's probably a bad idea"
- other ideas?

Personally I don't want poetry to manage my virtual environments at all, though I do want it to refuse to go ahead if I've forgotten to create the environment myself.  But probably it's hard to get to there from here; and certainly not everyone has the same workflow.

**Comment by kevr** [comment]
Please do not do this. This would seriously break Docker, and really, that config is up to the user; I don't see any problem with that, other than the fact that you run into users who choose to do the incorrect things on their systems.

**Comment by dimbleby** [comment]
The suggestion is not that it be made _impossible_ to use the system environment, people who insist on doing this can continue to do so.  Rather, that it be more clearly and firmly be discouraged.

The problem is that many of the people who avoid using a virtual environment don't really understand what they're doing, then things go wrong, then they come here... it's a big waste of everyone's time.  It would be better for those people to be guided towards a pattern that's more likely to work reliably in the first place.

(#6397 was the immediate prompt for opening this, but I'm confident there are lots of other examples)

**Comment by neersighted** [comment]
> Please do not do this. This would seriously break Docker workflows, and really, that config is up to the user; I don't see any problem with that, other than the fact that you run into users who choose to do the incorrect things on their systems.

I think you are among the users who might benefit from the documentation ask here -- there is no reason to avoid virtual environments in a container, and it usually causes more harm than good to do so.

Check out https://github.com/python-poetry/poetry/issues/6397#issuecomment-1236327500 for a recipe for 'seamless' use of a virtual environment in a container -- you get all the ergonomic benefits of not having to use a virtual environment path prefix, but you properly isolate your components so that you are not mixing dependency trees with unknown results.

**Comment by neersighted** [comment]
I wonder if splitting the option into `virtualenvs.manage` (so Poetry will fail if no env is active, but not create/directly manage the env) and `installer.system-site-packages` would make more sense in the long run.

**Comment by sullivancolin** [comment]
The documentation on the official installer seems to indicate that it will be installed in a  separate environment and therefore it would be safe to not use a virtualenv inside a  Docker container:

> We provide a custom installer that will install Poetry in a new virtual environment to isolate it from the rest of your system. This ensures that dependencies will not be accidentally upgraded or uninstalled, and allows Poetry to manage its own environment.


I can install poetry inside docker with these lines:
```dockerfile
RUN curl -sSL https://install.python-poetry.org | python3 - --version 1.2.0
ENV PATH "/home/vscode/.poetry/bin:$PATH"
RUN /home/vscode/.local/bin/poetry config virtualenvs.create false
```

When I then run `pip list` from system level python I get: 
```sh
Package    Version
---------- -------
pip        22.2.2
setuptools 58.1.0
wheel      0.37.1
```

This seems like it has been effectively isolated from the system level python.

However, once I run `poetry install` when `virtualenvs.create false` been set, the install fails.


Is the documentation wrong that the official installer is not actually isolated from system level python?  If not, what is the correct interpretation of what it's saying and why did installation succeed  with poetry<1.2?  If the documentation is wrong, then it does seem like stronger guardrails would be helpful.

**Comment by neersighted** [comment]
I wrote up a whole response but it looks like Github ate it -- sorry @sullivancolin.

The official installer, or any officially supported installation method will create a separate virtual environment to install Poetry into, separate from the system (or container, in this case -- we still call it system) level environment.

You can correctly see that no Poetry-related packages are installed into your system environment. However, `virtualenvs.create false` also forces your project to install into the system environment, which can be a bit fraught at best. Without specifics, I won't have specific insight of course, but in short 90% of users do not need `virtualenvs.create false` -- there's not a reason to not use a virtual environment in a container, and it protects you against a **vast** list of ways Python packaging can be a headache and go awry.

**Comment by neersighted** [comment]
> I wonder if splitting the option into `virtualenvs.manage` (so Poetry will fail if no env is active, but not create/directly manage the env) and `installer.system-site-packages` (or some similar split) would make more sense in the long run.
> 
> Honestly, I'd even add a warning if `installer.system-site-packages` is set with a docs link, since it's rarely something you actually want if you are fully informed as to how Python import paths and environments work.

This is tracked in https://github.com/python-poetry/poetry/issues/1945

**Comment by ewjoachim** [comment]
I think it would be useful if there was a way to tell procrastinate to never create a venv, but never use the system python either: to refuse to run if a venv is not set.

**Comment by neersighted** [comment]
This is a lot better with PEP 668 (https://peps.python.org/pep-0668/); it prevents the 'footgun' configurations where installing Poetry can conflict with e.g. DNF or APT. While it's still possible to have some footgun-like experiences with `library/python` container images, those are mostly from the much more obvious `pip install poetry; pip install something-else` and not the more subtle `pip install poetry; apt-get install <tool written in Python>`.

In other words, that is de-facto happening and users will benefit as they move to newer distros/Python versions; though we should probably verify that our installer also respects the PEP 668 markers.

**Comment by ZelphirKaltstahl** [comment]
Since various people are frequently eager to point out, that `poetry config virtualenvs.create false` is a bad idea, I would like to know, whether Poetry has already fixed the one main reason why people do this ... _naming your venvs and putting them in a place you determine, so that you can easily script usage of those venvs_. This is the one valid main motivation behind it, as far as I know.

So I ask you: What is your solution to this?

If there is none, then I think the advice _"is almost always a bad idea"_ without any reasoning is kinda patronizing, condescending or at the very least not understanding the _issue people have_ with how Poetry works.

I for one have used `poetry config virtualenvs.create false` without issues in docker builds and in named venv creation outside of docker, installing dependencies from a lock file. I cannot remember a single time, when that went wrong and did not do what I expected: installing the dependencies in the activated venv (if you have any activated). I don't want to not know the name of the venv in my containers have to chase down some hashed name somewhere. Ideally of course the venv is in the project directory itself. Even the default tool for creating venvs (`python3 -m venv myvenvname`) allows to specify location and name. It is only _reasonable_ to at least _initially_ expect a tool like Poetry to be able to do that.

I will note, that the plain statement _"is almost always a bad idea"_ or similar, without reasoning and _detailed_ explanation or example of what could go wrong, is likely to make people think that Poetry cannot do what they want it to do (named venv in decided location, not some silly hash name in not configured location), and makes it more likely people walk away to use other tools. (Which is entirely fine, but perhaps not what you want. Well, I don't know what you want, anyway.)

The very least one should do when advising people to not do that is to state, that the Poetry maintainers/developers/contributors have decided, that putting your venv where you want it is [not planned](https://github.com/python-poetry/poetry/issues/1579). That means, following your advice to not use the `poetry config virtualenvs.create false` configuration, it is hopeless, and people should look elsewhere. Please state this clearly, when you do the typical _"is almost always a bad idea"_ thing. Not stating this clearly is leading people on a wild goose chase to find any way to do what they want to do, before finally, doing `poetry config virtualenvs.create false` anyway, or annoyed switching to uv or something.

**Comment by dimbleby** [comment]
if you don't like the virtual environment that poetry creates, just create your own and have poetry use that
